internal/util/fieldSetter.go

```diff
@@ -42,70 +44,87 @@ func setParameter(serviceType interface{}, parameterHierarchy []string, value st
 		}
 
 		reflectedServiceType = reflectedServiceType.FieldByName(parameterName)
-
-		if reflectedServiceType.Kind() != reflect.Struct {
-			err := SetFields(reflectedServiceType, value)
-			if err != nil {
-				err := fmt.Errorf(
-					"%w\ncannot assign value %s to field %s with field Type %s",
-					err,
-					value,
-					strings.Join(parameterHierarchy, "."),
-					reflectedServiceType.FieldByName(parameterName).Kind(),
-				)
-				err = fmt.Errorf("%w\n%s contains field with name %s : %t",
-					err,
-					reflectedServiceType.Type().Name(),
-					parameterName,
-					reflectedServiceType.FieldByName(parameterName).IsValid(),
-				)
-				return nil, err
-			}
-			info := fmt.Sprintf("setting field: %s value: %s", strings.Join(parameterHierarchy, "."), value)
-			logrus.Info(info)
-		}
-
 	}
-	return serviceType, nil
-}
 
-// Returns all field names of a struct
-func getAllFieldNames(field reflect.Value) []string {
-	var fieldNames []string
-	for i := 0; i < field.NumField(); i++ {
-		if field.Type().Field(i).Anonymous {
-			// Recursively finds all field names of anonymous/embedded structs
-			fieldNames = append(fieldNames, getAllFieldNames(field.Field(i))...)
-		} else {
-			fieldNames = append(fieldNames, field.Type().Field(i).Name)
-		}
+	err = SetFields(reflectedServiceType, input)
+	if err != nil {
+		err = fmt.Errorf(
+			"%w cannot assign value %s to field %s with field Type %s",
+			err,
+			input.Value,
+			strings.Join(input.ParameterHierarchy, HIERARCHY_DELIMITER),
+			reflectedServiceType.Kind(),
+		)
+		err = fmt.Errorf("%w %s contains field with name %s : %t",
+			err,
+			reflectedServiceType.Type().Name(),
+			parameterName,
+			reflectedServiceType.IsValid(),
+		)
+		return nil, err
 	}
-	return fieldNames
+	info := fmt.Sprintf("setting field: %s value: %s", strings.Join(input.ParameterHierarchy, HIERARCHY_DELIMITER), input.Value)
+	logrus.Info(info)
+
+	return serviceType, nil
 }
 
 // Sets value of reflected field with type checking
-func SetFields(field reflect.Value, value string) error {
-	// TODO: Handle error cases on type conversion
-	if field.Kind() == reflect.String {
-		field.SetString(value)
+func SetFields(field reflect.Value, input Input) error {
+	if input.Unset {
+		field.Set(reflect.Zero(field.Type()))
+	} else if input.IsJson {
```

Comment:
This looks to me like a json argument is always parsed, there's no way to treat json as a string.

IMHO this part should be changed by reordering the various cases. Before handling the json case you should check for "field.Kind() == reflect.String" and then set the input as a string, even if the value is valid json.

Reply:
One issue with allowing json string values in this manner is that if you do something like:
`--spec.parameters.service.version {"Foo":"Bar"} --spec.parameters.service {"Baz":"Baas no"}` 
Now the ordering of the input matters. Since [].service is not of type string it will delete everything in [].service and only add `{"Foo":"Baar"}` if it is set after [].version, which is of type string. If the input order is reversed you would get the "correct" `{"Foo":"Bar","Baz":Baas"}`

Reply:
I'm not sure which use case takes precedence here.
Do we need the ablility to add json as a string ?

Reply:
> Now the ordering of the input matters.

I see no problem with that. I think the order matters anyway, e.g. you could have `--foo- --foo asdf` which would not be the same as `--foo asdf --foo-`

IMHO having a JSON string as a value something that should work. Yes it's a bit of a border case but not *that* uncommon. In the context of appcat this could easily happen if a back-end service takes JSON as configuration input (e.g. AWS network policies are JSON, so for instance it might happen that we have to configure an AWS RDS instance's network policy via a JSON string; or some system might have an "extra-settings" argument which might be a JSON string).

Reply:
Makes sense, thanks !
I've added the changes.